cmd/cody-gateway/internal/httpapi/attribution/handler.go

```diff
@@ -0,0 +1,75 @@
+package attribution
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/Khan/genqlient/graphql"
+
+	"github.com/sourcegraph/log"
+
+	"github.com/sourcegraph/sourcegraph/cmd/cody-gateway/internal/actor"
+	"github.com/sourcegraph/sourcegraph/internal/codygateway"
+	"github.com/sourcegraph/sourcegraph/internal/trace"
+)
+
+// Request for attribution search. Expected in JSON form as the body of POST request.
+type Request struct {
+	// Snippet is the text to search attribution of.
+	Snippet string
+	// Limit is the upper bound of number of responses we want to get.
+	Limit int
+}
+
+// Response of attribution search. Contains some repositories to which the snippet can be attributed to.
+type Response struct {
+	// Repositories which contain code matching search snippet.
+	Repositories []Repository
+	// TotalCount denotes how many total matches there were (including listed repositories).
+	TotalCount int
+	// LimitHit is true if the number of search hits goes beyond limit specified in request.
+	LimitHit bool
+}
+
+// Repository represents matching of search content against a repository.
+type Repository struct {
+	// Name of the repo on dotcom. Like github.com/sourcegraph/sourcegraph.
+	Name string
+}
+
+// NewHandler creates a REST handler for attribution search.
+// graphql.Client can be nil which disables the search.
+func NewHandler(client graphql.Client, baseLogger log.Logger) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		a := actor.FromContext(r.Context())
+		if got, want := a.GetSource(), codygateway.ActorSourceProductSubscription; got != want {
+			w.WriteHeader(http.StatusUnauthorized)
+			fmt.Fprintln(w, "only available for enterprise product subscriptions")
```

Comment:
We have a shared helper in Cody Gateway to render the error in a structured manner and generate debug/error logs based on status code:

```suggestion
            response.JSONError(logger, w, errors.New("only available for enterprise product subscriptions")
```